### ISSUE TITLE
Error when computing a cloned graph from xarray.open_dataset

### ISSUE THREAD

**Issue by tierriminator** [dask/dask#9621]
**Describe the issue**:

After opening a data array from xarray and cloning the underlying dask array, I receive a `TypeError: string indices must be integers` when running `compute()`.

**Minimal Complete Verifiable Example**:

```python
import dask.graph_manipulation
import numpy as np
import xarray as xr

if __name__ == "__main__":
    a = xr.DataArray(np.zeros((10, 10)), dims=("x", "y"), name="a")
    a.to_netcdf("a.nc")
    a: xr.DataArray = xr.open_dataarray("a.nc", chunks={"x": 5})
    c = dask.graph_manipulation.clone(a.data)
    c.compute()
```

**Anything else we need to know?**:

I was not sure whether to report the bug here or at the xarray repo. I hope it is fine like this.

After inspection with the debugger, I got the impression that the issue is due to a variable naming mismatch in the graph.
It seemed like `clone()` did not translate every occurrence of the old variable names. At least I've found an old variable name in the `Blockwise` construct.
In the original graph there was also a variable name containing dashes. So maybe it's a problem related to key grouping?

**Environment**:

I use a conda environment and the folloing versions:
```
# Name                    Version                   Build  Channel
dask                      2022.7.0        py310h06a4308_0  
xarray                    0.20.1             pyhd3eb1b0_1
numpy                     1.23.1          py310h1794996_0
```

**Comment by phobson** [comment]
Thanks for the reproducer. I see this locally with the more recent versions:

```
dask.__version__='2022.10.0'
np.__version__='1.23.4'
xr.__version__='2022.10.0'
```

<details><summary>Full trace back:</summary><pre>
2022-11-03 10:54:08,483 - distributed.worker - WARNING - Compute Failed
Key:       ('open_dataset-0fc1891630a83baf85e0b615c8b10441', 0, 0)
Function:  getter
args:      ('original-open_dataset-4f72518efde404bf84687164f381931ba-53dbb4e80dbf06565a8792dc89a57ed3', (slice(0, 5, None), slice(0, 10, None)))
kwargs:    {}
Exception: "TypeError('string indices must be integers')"

2022-11-03 10:54:08,484 - distributed.worker - WARNING - Compute Failed
Key:       ('open_dataset-0fc1891630a83baf85e0b615c8b10441', 1, 0)
Function:  getter
args:      ('original-open_dataset-4f72518efde404bf84687164f381931ba-53dbb4e80dbf06565a8792dc89a57ed3', (slice(5, 10, None), slice(0, 10, None)))
kwargs:    {}
Exception: "TypeError('string indices must be integers')"

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [67], line 9
      7 a: xr.DataArray = xr.open_dataarray("a.nc", chunks={"x": 5})
      8 c = dask.graph_manipulation.clone(a.data)
----> 9 c.compute()

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/dask/base.py:315, in DaskMethodsMixin.compute(self, **kwargs)
    291 def compute(self, **kwargs):
    292     """Compute this dask collection
    293 
    294     This turns a lazy Dask collection into its in-memory equivalent.
   (...)
    313     dask.base.compute
    314     """
--> 315     (result,) = compute(self, traverse=False, **kwargs)
    316     return result

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/dask/base.py:600, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    597     keys.append(x.__dask_keys__())
    598     postcomputes.append(x.__dask_postcompute__())
--> 600 results = schedule(dsk, keys, **kwargs)
    601 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/distributed/client.py:3057, in Client.get(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   3055         should_rejoin = False
   3056 try:
-> 3057     results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   3058 finally:
   3059     for f in futures.values():

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/distributed/client.py:2226, in Client.gather(self, futures, errors, direct, asynchronous)
   2224 else:
   2225     local_worker = None
-> 2226 return self.sync(
   2227     self._gather,
   2228     futures,
   2229     errors=errors,
   2230     direct=direct,
   2231     local_worker=local_worker,
   2232     asynchronous=asynchronous,
   2233 )

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/distributed/utils.py:339, in SyncMethodMixin.sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    337     return future
    338 else:
--> 339     return sync(
    340         self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    341     )

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/distributed/utils.py:406, in sync(loop, func, callback_timeout, *args, **kwargs)
    404 if error:
    405     typ, exc, tb = error
--> 406     raise exc.with_traceback(tb)
    407 else:
    408     return result

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/distributed/utils.py:379, in sync.<locals>.f()
    377         future = asyncio.wait_for(future, callback_timeout)
    378     future = asyncio.ensure_future(future)
--> 379     result = yield future
    380 except Exception:
    381     error = sys.exc_info()

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/tornado/gen.py:762, in Runner.run(self)
    759 exc_info = None
    761 try:
--> 762     value = future.result()
    763 except Exception:
    764     exc_info = sys.exc_info()

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/distributed/client.py:2089, in Client._gather(self, futures, errors, direct, local_worker)
   2087         exc = CancelledError(key)
   2088     else:
-> 2089         raise exception.with_traceback(traceback)
   2090     raise exc
   2091 if errors == "skip":

File ~/mambaforge/envs/dask-use/lib/python3.9/site-packages/dask/array/core.py:120, in getter()
    118     lock.acquire()
    119 try:
--> 120     c = a[b]
    121     # Below we special-case `np.matrix` to force a conversion to
    122     # `np.ndarray` and preserve original Dask behavior for `getter`,
    123     # as for all purposes `np.matrix` is array-like and thus
    124     # `is_arraylike` evaluates to `True` in that case.
    125     if asarray and (not is_arraylike(c) or isinstance(c, np.matrix)):

TypeError: string indices must be integers
</pre></details>

**Comment by CSSFrancis** [comment]
Is there any movement on this? I recently ran into the same bug and was wondering if there was a work around in the mean time?

**Comment by CSSFrancis** [comment]
Just a note that the following code fails as well.  It seems like there is a key which gets unnecessarily trimmed during a graph optimization call


```python
import dask.array as da
import numpy as np
from dask.graph_manipulation import clone

d = np.ones((10,10))
fp = np.memmap("test", dtype='float64', mode='w+', shape=(10, 10))
fp[:,:]=d
fp.flush()
fp = np.memmap("test", dtype='float64', mode='r+', shape=(10, 10))
s = da.from_array(fp)
clone(s).compute()
```

**Comment by jrbourbeau** [comment]
Thanks @CSSFrancis -- I'm not aware of any developments. cc @crusaderky for the `dask.graph_manipulation` and `xarray` connection

**Comment by CSSFrancis** [comment]
Just for completeness sake, the following code throws the same error as well.

```
import zarr
import dask.array as da
from dask.graph_manipulation import clone
arr = zarr.ones((10,10))
d = da.from_zarr(arr)
clone(d).compute()
````

**Comment by crusaderky** [comment]
Barebones reproducer:
```python
import operator
import numpy
import dask.array as da
from dask.graph_manipulation import clone
from dask.highlevelgraph import HighLevelGraph

hlg = HighLevelGraph(
    layers={
        "x-123": {"x-123": (numpy.arange, 10)},
        "y-456": {("y-456", 0): "x-123"},
    },
    dependencies={
        "y-456": {"x-123"},
    },
)
a = da.Array(hlg, name="y-456", shape=(3, ), chunks=((3, ),), dtype=int, meta=numpy.arange(0))
a.compute()
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
clone(a).compute()
KeyError: 'x-123'
```

**Comment by crusaderky** [comment]
The problem is that the original dask graph is missing an empty `dependencies` entry for `x-123`.

**Comment by crusaderky** [comment]
The problem is that the original dask graph is missing an empty `dependencies` entry for `x-123`.

**Comment by crusaderky** [comment]
This is a regression introduced in #8979. If I revert it, this is fixed, but #8978 gets broken again. Busy finding a solution that works for both.

**Comment by crusaderky** [comment]
PR ready for review
